hack/tools/janitor/janitor.go

```diff
@@ -0,0 +1,245 @@
+	managedObjects, err := finder.ManagedObjectListChildren(ctx, folder+"/...", "VirtualMachine")
+	if err != nil {
+		return err
+	}
+
+	if len(managedObjects) == 0 {
+		return nil
+	}
+
+	// Retrieve information for all found virtual machines.
+	managedObjectReferences := []types.ManagedObjectReference{}
+	for _, obj := range managedObjects {
+		managedObjectReferences = append(managedObjectReferences, obj.Object.Reference())
+	}
+	var managedObjectVMs []mo.VirtualMachine
+	if err := s.vSphereClients.Govmomi.Retrieve(ctx, managedObjectReferences, []string{"config", "summary.runtime.powerState"}, &managedObjectVMs); err != nil {
+		return err
+	}
+
+	vmsToDeleteAndPoweroff := []*virtualMachine{}
+	vmsToDelete := []*virtualMachine{}
+
+	// Filter out vms we don't have to cleanup depending on s.maxCreationDate.
+	for _, managedObjectVM := range managedObjectVMs {
+		if managedObjectVM.Config.CreateDate.After(s.maxCreationDate) {
+			// Ignore vms created after maxCreationDate
+			continue
+		}
+
+		vm := &virtualMachine{
+			managedObject: managedObjectVM,
+			object:        object.NewVirtualMachine(s.vSphereClients.Vim, managedObjectVM.Reference()),
+		}
+
+		if vm.managedObject.Summary.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
+			vmsToDeleteAndPoweroff = append(vmsToDeleteAndPoweroff, vm)
+			continue
+		}
+		vmsToDelete = append(vmsToDelete, vm)
+	}
+
+	// PowerOff vms which are still running. Triggering PowerOff for a VM results in a task in vSphere.
+	poweroffTasks := []*object.Task{}
+	for _, vm := range vmsToDeleteAndPoweroff {
+		log.Info("Powering off vm in vSphere", "vm", vm.managedObject.Config.Name)
+		if s.dryRun {
+			// Skipping actual PowerOff on dryRun.
+			continue
+		}
+		task, err := vm.object.PowerOff(ctx)
+		if err != nil {
+			return err
+		}
+		poweroffTasks = append(poweroffTasks, task)
+	}
+	// Wait for all PowerOff tasks to be finished. We intentionally ignore errors here
+	// because the VM may already got into PowerOff state and log the errors only.
+	// xref govc: https://github.com/vmware/govmomi/blob/512c168/govc/vm/destroy.go#L94-L96
+	if err := waitForTasksFinished(ctx, poweroffTasks, true); err != nil {
+		log.Info("Ignoring error for PowerOff task", "err", err)
+	}
+
+	destroyTasks := []*object.Task{}
+	for _, vm := range append(vmsToDeleteAndPoweroff, vmsToDelete...) {
+		log.Info("Destroying vm in vSphere", "vm", vm.managedObject.Config.Name)
+		if dryRun {
+			// Skipping actual destroy on dryRun.
+			continue
+		}
+		task, err := vm.object.Destroy(ctx)
+		if err != nil {
+			return err
+		}
+		destroyTasks = append(destroyTasks, task)
+	}
+	// Wait for all destroy tasks to succeed.
+	if err := waitForTasksFinished(ctx, destroyTasks, false); err != nil {
+		return errors.Wrap(err, "failed to wait for vm poweroff task to finish")
+	}
+
+	return nil
+}
+
+func waitForTasksFinished(ctx context.Context, tasks []*object.Task, ignoreErrors bool) error {
```

Comment:
Would it make sense to print all tasks somewhere? So if we already return after the first failed we have a list of taskIDs we can use to look into the problem?